src/coreclr/vm/codeman.cpp

```diff
@@ -1525,6 +1528,18 @@ void EEJitManager::SetCpuInfo()
 
     if (((cpuFeatures & ARM64IntrinsicConstants_Sve) != 0) && CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnableArm64Sve))
     {
+#if defined(TARGET_LINUX) && (defined(DEBUG) || defined(_DEBUG))
```

Comment:
Why is this restricted to DEBUG only?

I also expect that we need a path for Windows and should likely treat SVE as unsupported if the vector length is larger and restricting the size (via `prctl` or equivalent on other OS) fails.

Reply:
This option is added for development purpose only.  It's primarily added to aid implementation of API, and its testing on a 256bit SVE enabled V1 system that we have access to. When Vector<T> would use the entire vector length, it may become redundant and be removed.

Reply:
I don't think it being for development purposes only is a good thing long term and it probably doesn't match up with the intent of `DOTNET_MaxVectorTBitWidth`.

We don't really want to disable SVE on hardware with 256-bit vectors just because a user has said they want 128-bit vectors, using the OS feature like `prctl` should be a much better option so that they can still get SVE usage in that scenario.

Reply:
> I also expect that we need a path for Windows and should likely treat SVE as unsupported if the vector length is larger and restricting the size (via `prctl` or equivalent on other OS) fails.

I will check with Windows team on what is the equivalent API.

Reply:
Kunal asked me to share how this works in Windows (I developed the SVE support in Windows for the next release). We felt that being able to dynamically change the vector length at runtime would be messy, since there's no save/restore mechanism for these vector length changes, hence you could get into situations where for example some code calls into a library, the library decreases the VL for some reason, but then has a bug where under certain conditions it fails to restore the vector length, so after the caller gains control again it would be running with a decreased vector length indefinitely, etc.

We do recognize the need to change the vector length for various purposes (testing, perf, compat, etc), so we did add a CreateProcess parameter that allows the vector length to be specified during process creation, but after a process has been created, its vector length cannot be changed. There's also registry keys that can be set to change the vector length on a per-process basis (IFEO settings), or for all processes in the system, but these registry-based methods are typically used for development and testing only.

By default, we'll use the highest VL supported by the system and the underlying hypervisor. So the above mechanisms will only be necessary when you want to decrease the VL for processes.

Happy to explore this topic more with you.

Reply:
Thanks for the context @JasonLinMS. 

I definitely agree that changing this during the lifetime of the process is messy/problematic and the intent in general is to not allow that for .NET either.

I think it's probably worth us having a short meeting (.NET, Windows, and Arm) to see if we can discuss things here and see if we can find something that generally works for everyone. CC. @jkotas 

-----------------

In the case of .NET, we would really only need the capability to set this once as part of our own startup before any user code has executed. We don't have any intent to change this dynamically (although there is the SME feature that makes this a little bit more complicated) and the ideal for user code is to write size agnostic algorithms so that it doesn't matter what size the hardware actually supports.

However, the official Arm64 SVE/Vector ABIs (https://github.com/ARM-software/abi-aa?tab=readme-ov-file#abi-for-the-arm-64-bit-architecture-with-sve-support) do define the ability to say a given API expects a particular size and there may be cases where a user needs to fix it themselves (potentially just for testing purposes or giving users a workaround for a bug). As such, the ability for a process to set the size for itself is beneficial, especially if that can be shared across all Arm64 capable operating systems.

Reply:
Yes, a meeting to discuss this sounds good.

Reply:
scheduled.

Reply:
Once other feedback is addressed, this change can go in without having us to wait for Windows support. 

I think this should also check the other requirements such as it should by 128-bit increments only and maximum should be 2048 and such.

Reply:
> we did add a CreateProcess parameter that allows the vector length to be specified during process creation, but after a process has been created, its vector length cannot be changed.

This is a very sensible choice. In Linux, changing the vector length inside a library and not restoring it before returning is generally seen as undefined behaviour. OpenJDK doesn't want to trust that, and after calling external routines it inserts checks to confirm the vector length hasn't changed. Neither of which is ideal.

> However, the official Arm64 SVE/Vector ABIs (https://github.com/ARM-software/abi-aa?tab=readme-ov-file#abi-for-the-arm-64-bit-architecture-with-sve-support) do define the ability to say a given API expects a particular size

The PCS isn't clear if the vector length must remain fixed. I've raised a bug against it [here](https://github.com/ARM-software/abi-aa/issues/260). It would be good to have a clear statement.



For coreclr, if changing the vector length is only for debugging then a wrapper script/binary that just launches coreclr with the correct vector length might be good enough. That would work for Windows and Linux. Ideally I'd still like to keep this PRs mechanism for use in Linux.